support/metrics/sets.go

```diff
@@ -0,0 +1,329 @@
+package metrics
+
+import (
+	"os"
+	"strings"
+
+	prometheusoperatorv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type MetricsSet string
+
+const (
+	MetricsSetTelemetry MetricsSet = "Telemetry"
+	MetricsSetSRE       MetricsSet = "SRE"
+	MetricsSetAll       MetricsSet = "All"
+)
+
+const (
+	metricsSetEnvVar  = "METRICS_SET"
+	DefaultMetricsSet = MetricsSetTelemetry
+)
+
+var (
+	allMetricsSets = sets.NewString()
+)
+
+func init() {
+	allMetricsSets.Insert(string(MetricsSetTelemetry))
+	allMetricsSets.Insert(string(MetricsSetSRE))
+	allMetricsSets.Insert(string(MetricsSetAll))
+}
+
+func MetricsSetFromString(str string) MetricsSet {
+	if str == "" {
+		return DefaultMetricsSet
+	}
+	for _, value := range allMetricsSets.List() {
+		if strings.EqualFold(string(value), str) {
+			return MetricsSet(value)
+		}
+	}
+	return DefaultMetricsSet
+}
+
+func MetricsSetFromEnv() MetricsSet {
+	return MetricsSetFromString(os.Getenv(metricsSetEnvVar))
+}
+
+func MetricsSetToEnv(set MetricsSet) corev1.EnvVar {
+	return corev1.EnvVar{
+		Name:  metricsSetEnvVar,
+		Value: string(set),
+	}
+}
+
+func (s *MetricsSet) String() string {
+	return string(*s)
+}
+
+func (s *MetricsSet) Set(value string) error {
+	*s = MetricsSetFromString(value)
+	return nil
+}
+
+func (s *MetricsSet) Type() string {
+	return "metricsSet"
+}
+
+func CVORelabelConfigs(set MetricsSet) []*prometheusoperatorv1.RelabelConfig {
+	switch set {
+	case MetricsSetTelemetry:
+		return []*prometheusoperatorv1.RelabelConfig{
+			{
+				Action:       "keep",
+				Regex:        "(cluster_version|cluster_version_available_updates|cluster_operator_up|cluster_operator_conditions|cluster_version_payload|cluster_installer)",
+				SourceLabels: []string{"__name__"},
+			},
+		}
+
```

Comment:
i might be misunderstanding something here. If someone used the `SRE` or `All` metric sets, wouldn't the expectation be that those are supersets of the basic `Telemetry` set?

Reply:
They are. The relabel rule for telemetry is a whitelist of the metrics allowed. The rules for the SRE and All sets are metrics that are dropped.